File: algorithms/insertion_sort/python/insertion_sort.py
#!/usr/bin/env python3

from typing import List

def insertion_sort(arr: List, size: int) -> None:
  for i in range(1, size):
    j = i
    key = arr[i]

    while j > 0 and arr[j - 1] > key:
      arr[j] = arr[j - 1]
      j -= 1

    arr[j] = key

if __name__ == "__main__":
  arr = [6, 3, 5, 10, -1, 7, -4]
  insertion_sort(arr, len(arr))

File: algorithms/insertion-sort/cc/insertion_sort.cc
#include <iostream>

template <typename T>
void insertion_sort(T arr[], int size)
{
  for (int i = 1; i < size; i++) {
    T key = arr[i];
    int j = i - 1;

    while (j >= 0 && arr[j] > key) {
      arr[j + 1] = arr[j];
      j--;
    }

    arr[j + 1] = key;
  }
}

int main()
{
  int arr[] = {5, 2, 4, 6, 1, 3};
  insertion_sort<int>(arr, 6);
  return 0;
}

